### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by tpunt**
Hi Joe!

I've done a little refactoring on parts of pthreads. Nothing seems to be broken by these changes, which focus on removing redundant logic and centralising logic to prevent repetition. All changes are self-contained in individual commits, so if there are any changes you disagree with, then the others can simply be cherry-picked instead.

### CODE REVIEW

**Comment by krakjoe** [comment]
At a glance, looks good ... will review properly later on today ...

Thanks :)

File: src/prepare.c [hunk 1]
```diff
@@ -544,9 +544,7 @@ static inline void pthreads_prepare_functions(pthreads_object_t* thread) {
 			continue;
 		}
 
-		if (!(GC_FLAGS(name) & IS_STR_PERSISTENT)) {
```
Comment (krakjoe) [inline src/prepare.c]: This is left overs from pthreads_global_string, which is no longer used.

You may remove that function from globals.* to complete the job there ;)

Comment (tpunt) [inline src/prepare.c]: Done (b4da9e316ec8d78431a3fd902c988db4149a3adc) :)

**Comment by krakjoe** [comment]
Can you build in debug mode and have a look at this:

```
=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
Testing class statics properties (gh bug #570) [tests/class-static-properties.phpt]
=====================================================================
```

```
---- EXPECTED OUTPUT
string(3) "one"
array(3) {
  [0]=>
  int(1)
  [1]=>
  int(2)
  [2]=>
  int(3)
}
array(2) {
  [0]=>
  array(3) {
    [0]=>
    int(1)
    [1]=>
    int(2)
    [2]=>
    int(3)
  }
  [1]=>
  array(3) {
    [0]=>
    int(4)
    [1]=>
    int(5)
    [2]=>
    int(6)
  }
}
object(Threaded)#1 (0) {
}
NULL
object(Closure)#2 (0) {
}
array(3) {
  [0]=>
  int(1)
  [1]=>
  int(2)
  [2]=>
  int(3)
}
NULL
array(0) {
}
array(2) {
  [0]=>
  object(Threaded)#3 (0) {
  }
  [2]=>
  array(2) {
    [0]=>
    object(Threaded)#4 (0) {
    }
    [2]=>
    array(1) {
      [0]=>
      object(Threaded)#5 (0) {
      }
    }
  }
}
---- ACTUAL OUTPUT
string(3) "one"
array(3) {
  [0]=>
  int(1)
  [1]=>
  int(2)
  [2]=>
  int(3)
}
array(2) {
  [0]=>
  array(3) {
    [0]=>
    int(1)
    [1]=>
    int(2)
    [2]=>
    int(3)
  }
  [1]=>
  array(3) {
    [0]=>
    int(4)
    [1]=>
    int(5)
    [2]=>
    int(6)
  }
}
object(Threaded)#1 (0) {
}
NULL
object(Closure)#2 (0) {
}
array(3) {
  [0]=>
  int(1)
  [1]=>
  int(2)
  [2]=>
  int(3)
}
NULL
array(0) {
}
array(2) {
  [0]=>
  object(Threaded)#3 (0) {
  }
  [2]=>
  array(2) {
    [0]=>
    object(Threaded)#4 (0) {
    }
    [2]=>
    array(1) {
      [0]=>
      object(Threaded)#5 (0) {
      }
    }
  }
}
[Thu Nov 10 16:57:15 2016]  Script:  '/usr/src/pthreads/tests/class-static-properties.php'
/usr/src/php-src/Zend/zend_API.c(1059) :  Freeing 0x2B7B888023C0 (56 bytes), script=/usr/src/pthreads/tests/class-static-properties.php
Last leak repeated 1 time
[Thu Nov 10 16:57:15 2016]  Script:  '/usr/src/pthreads/tests/class-static-properties.php'
/usr/src/php-src/Zend/zend_vm_execute.h(19624) :  Freeing 0x2B7B88802660 (56 bytes), script=/usr/src/pthreads/tests/class-static-properties.php
Last leak repeated 1 time
[Thu Nov 10 16:57:15 2016]  Script:  '/usr/src/pthreads/tests/class-static-properties.php'
/usr/src/php-src/Zend/zend_hash.c(135) :  Freeing 0x2B7B88860DC0 (264 bytes), script=/usr/src/pthreads/tests/class-static-properties.php
Last leak repeated 3 times
[Thu Nov 10 16:57:15 2016]  Script:  '/usr/src/pthreads/tests/class-static-properties.php'
/usr/src/php-src/Zend/zend_objects.c(162) :  Freeing 0x2B7B88865CD0 (40 bytes), script=/usr/src/pthreads/tests/class-static-properties.php
Last leak repeated 1 time
[Thu Nov 10 16:57:15 2016]  Script:  '/usr/src/pthreads/tests/class-static-properties.php'
/usr/src/pthreads/src/globals.c(77) :  Freeing 0x2B7B8887A280 (128 bytes), script=/usr/src/pthreads/tests/class-static-properties.php
/usr/src/php-src/Zend/zend_alloc.c(2505) : Actual location (location was relayed)
Last leak repeated 1 time
```

**Comment by tpunt** [comment]
I didn't have time to look into this tonight, but I should have some time tomorrow morning to (if you don't beat me to it) :)

**Comment by tpunt** [comment]
Regarding the memory leak, what's wrong with simply decrementing `pzval`'s refcount by one in `pthreads_store_remove_complex`?
```
--GC_REFCOUNT(Z_COUNTED_P(pzval));
```

**Comment by krakjoe** [comment]
Was this leak already here ?

**Pull request merged**